pages/blog/posts/understanding-lexical-dynamic-scopes.md

```diff
@@ -0,0 +1,412 @@
+On the right, a directed graph representation of the relationship between these
+schema resources and the dynamic scope.
+
+As usual, the evaluation process starts with the top level schema. The dynamic
+scope at that point is the root schema resource, and all other schema resources
+are out of scope:
+
+![The dynamic scope as a stack (1)](/img/posts/2024/understanding-lexical-dynamic-scopes/dynamic-scope-pop-1.webp)
+
+Next, we enter the `if` applicator that checks whether the integer instance is
+even or odd. This subschema declares a new schema resource called
+`https://example.com/check`, which is pushed onto the stack. Therefore the
+dynamic scope consists of `https://example.com/integer` followed by
+`https://example.com/check`:
+
+![The dynamic scope as a stack (2)](/img/posts/2024/understanding-lexical-dynamic-scopes/dynamic-scope-pop-2.webp)
+
+The `https://example.com/check` nested schema resource does not reference any
+other schema resource. When the evaluation process completes and determines
+that the instance is an even integer, the stack unwinds, the
+`https://example.com/check` schema resource is popped, and the evaluation
+process *returns* to the root schema resource. Therefore the dynamic scope is
+back to just `https://example.com/integer`:
+
+![The dynamic scope as a stack (3)](/img/posts/2024/understanding-lexical-dynamic-scopes/dynamic-scope-pop-3.webp)
+
+Because the `if` subschema successfully validated the instance, we enter the
+`then` applicator. This subschema declares a new schema resource called
+`https://example.com/even`, which is pushed onto the stack. Therefore the
+dynamic scope consists of `https://example.com/integer` followed by
+`https://example.com/even`:
+
+![The dynamic scope as a stack (4)](/img/posts/2024/understanding-lexical-dynamic-scopes/dynamic-scope-pop-4.webp)
+
+Like before, the `https://example.com/even` nested schema resource does not
+reference any other schema resource. Therefore, the evaluation process returns
+once more to the root schema resource, the dynamic scope is back to just
+`https://example.com/integer`, and the evaluation process completes:
+
+![The dynamic scope as a stack (5)](/img/posts/2024/understanding-lexical-dynamic-scopes/dynamic-scope-pop-5.webp)
+
+Summary
+-------
+
+Understanding how static and dynamic scopes work is essential for gaining a
+deeper understanding of JSON Schema. The most important points to remember are
+summarized in the following table:
+
+| Comparison Point                     | Lexical Scope | Dynamic Scope |
+|--------------------------------------|---------------|---------------|
+| **Definition**                       | Consists of the schema resource being evaluated | Consists of the stack of schema resources evaluated so far |
+| **Determining the scope**            | Can be statically determined without taking instances into account | Cannot be always statically determined. It may vary depending on the instance |
```

Comment:
```suggestion
| **Determining the scope**            | Can be statically determined without taking instances into account | Cannot always be statically determined. It may vary depending on the instance |
```